In [65]:
# importing libraries
import pandas as pd
import numpy as np

In [66]:
# importing dataset
df = pd.read_csv('Melbourne_housing_FULL.csv')
df.head()

Suburb             Address  Rooms Type      Price Method SellerG  \
0  Abbotsford       68 Studley St      2    h        NaN     SS  Jellis   
1  Abbotsford        85 Turner St      2    h  1480000.0      S  Biggin   
2  Abbotsford     25 Bloomburg St      2    h  1035000.0      S  Biggin   
3  Abbotsford  18/659 Victoria St      3    u        NaN     VB  Rounds   
4  Abbotsford        5 Charles St      3    h  1465000.0     SP  Biggin   

        Date  Distance  Postcode  ...  Bathroom  Car  Landsize  BuildingArea  \
0  3/09/2016       2.5    3067.0  ...       1.0  1.0     126.0           NaN   
1  3/12/2016       2.5    3067.0  ...       1.0  1.0     202.0           NaN   
2  4/02/2016       2.5    3067.0  ...       1.0  0.0     156.0          79.0   
3  4/02/2016       2.5    3067.0  ...       2.0  1.0       0.0           NaN   
4  4/03/2017       2.5    3067.0  ...       2.0  0.0     134.0         150.0   

   YearBuilt         CouncilArea Lattitude  Longtitude             Regionname  \
0        NaN  Yarra City Council  -37.8014    144.9958  Northern Metropolitan   
1        NaN  Yarra City Council  -37.7996    144.9984  Northern Metropolitan   
2     1900.0  Yarra City Council  -37.8079    144.9934  Northern Metropolitan   
3        NaN  Yarra City Council  -37.8114    145.0116  Northern Metropolitan   
4     1900.0  Yarra City Council  -37.8093    144.9944  Northern Metropolitan   

  Propertycount  
0        4019.0  
1        4019.0  
2        4019.0  
3        4019.0  
4        4019.0  

[5 rows x 21 columns]

In [67]:
# checking shape of the dataset
df.shape

(34857, 21)

In [68]:
# checking names of columns of the dataset
df.columns

Index(['Suburb', 'Address', 'Rooms', 'Type', 'Price', 'Method', 'SellerG',
       'Date', 'Distance', 'Postcode', 'Bedroom2', 'Bathroom', 'Car',
       'Landsize', 'BuildingArea', 'YearBuilt', 'CouncilArea', 'Lattitude',
       'Longtitude', 'Regionname', 'Propertycount'],
      dtype='object')

In [69]:
# retaining useful features/columns for data analysis
useful_col = ['Suburb','Rooms', 'Type', 'Method', 'SellerG', 'Distance', 'Bedroom2', 'Bathroom', 'Car', 
            'Landsize', 'BuildingArea', 'CouncilArea', 'Regionname', 'Propertycount', 'Price']

df = df[useful_col]
df.head()

Suburb  Rooms Type Method SellerG  Distance  Bedroom2  Bathroom  Car  \
0  Abbotsford      2    h     SS  Jellis       2.5       2.0       1.0  1.0   
1  Abbotsford      2    h      S  Biggin       2.5       2.0       1.0  1.0   
2  Abbotsford      2    h      S  Biggin       2.5       2.0       1.0  0.0   
3  Abbotsford      3    u     VB  Rounds       2.5       3.0       2.0  1.0   
4  Abbotsford      3    h     SP  Biggin       2.5       3.0       2.0  0.0   

   Landsize  BuildingArea         CouncilArea             Regionname  \
0     126.0           NaN  Yarra City Council  Northern Metropolitan   
1     202.0           NaN  Yarra City Council  Northern Metropolitan   
2     156.0          79.0  Yarra City Council  Northern Metropolitan   
3       0.0           NaN  Yarra City Council  Northern Metropolitan   
4     134.0         150.0  Yarra City Council  Northern Metropolitan   

   Propertycount      Price  
0         4019.0        NaN  
1         4019.0  1480000.0  
2         4019.0  1035000.0  
3         4019.0        NaN  
4         4019.0  1465000.0

In [70]:
# checking again the shape of the dataset
df.shape

(34857, 15)

In [71]:
# checking Null values of the dataset
df.isnull().sum()

Suburb               0
Rooms                0
Type                 0
Method               0
SellerG              0
Distance             1
Bedroom2          8217
Bathroom          8226
Car               8728
Landsize         11810
BuildingArea     21115
CouncilArea          3
Regionname           3
Propertycount        3
Price             7610
dtype: int64

In [72]:
# filling null values with '0' in few columns
miss_val = ['Distance', 'Bedroom2', 'Bathroom', 'Car','Propertycount']
df[miss_val] = df[miss_val].fillna(0)

In [73]:
# checking Null values of the dataset again
df.isnull().sum()

Suburb               0
Rooms                0
Type                 0
Method               0
SellerG              0
Distance             0
Bedroom2             0
Bathroom             0
Car                  0
Landsize         11810
BuildingArea     21115
CouncilArea          3
Regionname           3
Propertycount        0
Price             7610
dtype: int64

In [74]:
# filling null values with 'mean' in 02 columns
df['Landsize'] = df['Landsize'].fillna(df.Landsize.mean())
df['BuildingArea'] = df['BuildingArea'].fillna(df.BuildingArea.mean())

In [75]:
# checking Null values of the dataset again
df.isnull().sum()

Suburb              0
Rooms               0
Type                0
Method              0
SellerG             0
Distance            0
Bedroom2            0
Bathroom            0
Car                 0
Landsize            0
BuildingArea        0
CouncilArea         3
Regionname          3
Propertycount       0
Price            7610
dtype: int64

In [76]:
# dropping remaining Null values
df.dropna(inplace=True)
df.isnull().sum()

Suburb           0
Rooms            0
Type             0
Method           0
SellerG          0
Distance         0
Bedroom2         0
Bathroom         0
Car              0
Landsize         0
BuildingArea     0
CouncilArea      0
Regionname       0
Propertycount    0
Price            0
dtype: int64

In [77]:
# treating categorical columns of the dataset through dummies variables
df = pd.get_dummies(df, drop_first=True) # drop_first = True to avoid dummy trap
df.head()

Rooms  Distance  Bedroom2  Bathroom  Car  Landsize  BuildingArea  \
1      2       2.5       2.0       1.0  1.0     202.0      160.2564   
2      2       2.5       2.0       1.0  0.0     156.0       79.0000   
4      3       2.5       3.0       2.0  0.0     134.0      150.0000   
5      3       2.5       3.0       2.0  1.0      94.0      160.2564   
6      4       2.5       3.0       1.0  2.0     120.0      142.0000   

   Propertycount      Price  Suburb_Aberfeldie  ...  \
1         4019.0  1480000.0                  0  ...   
2         4019.0  1035000.0                  0  ...   
4         4019.0  1465000.0                  0  ...   
5         4019.0   850000.0                  0  ...   
6         4019.0  1600000.0                  0  ...   

   CouncilArea_Wyndham City Council  CouncilArea_Yarra City Council  \
1                                 0                               1   
2                                 0                               1   
4                                 0                               1   
5                                 0                               1   
6                                 0                               1   

   CouncilArea_Yarra Ranges Shire Council  Regionname_Eastern Victoria  \
1                                       0                            0   
2                                       0                            0   
4                                       0                            0   
5                                       0                            0   
6                                       0                            0   

   Regionname_Northern Metropolitan  Regionname_Northern Victoria  \
1                                 1                             0   
2                                 1                             0   
4                                 1                             0   
5                                 1                             0   
6                                 1                             0   

   Regionname_South-Eastern Metropolitan  Regionname_Southern Metropolitan  \
1                                      0                                 0   
2                                      0                                 0   
4                                      0                                 0   
5                                      0                                 0   
6                                      0                                 0   

   Regionname_Western Metropolitan  Regionname_Western Victoria  
1                                0                            0  
2                                0                            0  
4                                0                            0  
5                                0                            0  
6                                0                            0  

[5 rows x 745 columns]

In [78]:
# obtaining features and labels
X = df.drop('Price', axis=1)
y = df['Price']

In [79]:
# splitting data into train and test data
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=2)

## The Problem of Multicollinearity/High Variance/OverFitting in Regression
One problem that often occurs in practice with multiple linear regression is multicollinearity – when two or more predictor variables are highly correlated to each other, such that they do not provide unique or independent information in the regression model.

This can cause the coefficient estimates of the model to be unreliable and have high variance (overfitting). That is, when the model is applied to a new set of data it hasn’t seen before, it’s likely to perform poorly.
Below mentioned model is performing reasonably well on training dataset and poorly performing on training dataset.

In [80]:
from sklearn.linear_model import LinearRegression
lr = LinearRegression()
lr.fit(X_train, y_train)
print('Simple Regression Test Score:', lr.score(X_test, y_test))
print('Simple Regression Train Score:', lr.score(X_train, y_train))

Simple Regression Test Score: 0.1385368316153236
Simple Regression Train Score: 0.6827792395792723


In this case accuracy on test dataset is just 13.85% as compared to training dataset i.e. 68.27%.

## Avoiding Multicollinearity/High Variance/OverFitting: 
If there is a large number of features in a dataset, we could easily end up overfitting the model by poorly selecting the features.
We can use regularization to get around this issue of Multicollinearity/High Variance/OverFitting using ridge regression and lasso regression.

## Regularization
### Regularization is a technique to optimize model performance and it does so by adding a small bias in the cost function. This small bias shrinks feature coefficients and reduces their sensitivity.

### What a cost function is?
Y = β0 + β1X1 + β2X2 + … + βpXp + ε
The values for β0, β1, β2, … , βp are chosen using the least square method, which minimizes the sum of squared residuals (RSS):

RSS = Σ(yi – ŷi)2

where:
Σ: A symbol that means “sum”
yi: The actual response value for the ith observation
ŷi: The predicted response value for the ith observation
The difference between a predicted and an actual value is called the error. Every data point in a dataset creates such errors, and the role of a cost function is to quantify these errors. And in linear regression, the objective is to minimize those errors to find the best fit model.

There are several cost functions out there such as Mean Squared Error (MSE), Mean Absolute Error (MAE), sum of squared residuals (RSS, Root Mean Squared Error (RMSE) etc. but they all do the same thing — quantifying errors.
It is the value of cost function that we want to minimize to find the best model.

### Bias and variance trade-off
Variance is always associated with a model failing to perform well on a new (test) set. High variance leads to overfitting which makes sense because overfit models generate vastly different, inconsistent results for different datasets.

In contrast, bias is the inability of the model to generalize on the training data. A model with too much bias fails to generalize both on training and test sets.

Ideally, the perfect model would have low bias and low variance but that is easier said than done. Because, bias and variance are in a trade-off relationship over model complexity.

### How Regularization works: 
Two types of regularization are used — L1 (LASSO regression) and L2 (Ridge regression). They are controlled by a hyperparameter called λ (α is used instead of λ in Sklearn library).

Ridge regression seeks to minimize the following:
#### - RSS + λΣβj2
Lasso regression seeks to minimize the following:
#### - RSS + λΣ|βj|
In both equations, the second term is known as a shrinkage penalty.
- When λ = 0, this penalty term has no effect and both ridge regression and lasso regression produce the same coefficient estimates as least squares.
- However, as λ approaches infinity the shrinkage penalty becomes more influential and the predictor variables that aren’t importable in the model get shrunk towards zero.
- The benefit of ridge and lasso regression compared to least squares regression lies in the bias-variance tradeoff.
- The basic idea of both ridge and lasso regression is to introduce a little bias so that the variance can be substantially reduced, which leads to a lower overall RSS.
Let's discuss one by one.

### Ridge Regression

Ridge
Cost function = Loss + λ x∑‖w‖^2 
where,
Loss = Sum of the squared residuals (SSE), λ = Penalty for the errors, w = slope of the curve/line
Ridge regression performs ‘L2 regularization‘, i.e. it adds a factor of sum of squares of coefficients in the optimization objective.
The purpose of L2 is to shrink feature coefficients to close to zero, but not exactly zero. Thus, ridge regression optimizes the following:

Objective = RSS + α * (sum of square of coefficients)
Here, α (alpha) is the parameter which balances the amount of emphasis given to minimizing RSS vs minimizing sum of square of coefficients. α can take various values:

#### α = 0:
The objective becomes same as simple linear regression. We’ll get the same coefficients as simple linear regression.
#### α = ∞:
The coefficients will be zero. Why? Because of infinite weightage on square of coefficients, anything less than zero will make the objective infinite.
#### 0 < α < ∞:
The magnitude of α will decide the weightage given to different parts of objective.

In [81]:
# applying ridge regression using alpha value as 1.0
from sklearn.linear_model import Ridge
l2 = Ridge(alpha=1)
l2.fit(X_train, y_train)
print('Ridge Test Score:',l2.score(X_test, y_test))
print('Ridge Train Score:',l2.score(X_train, y_train))

Ridge Test Score: 0.6701765758295284
Ridge Train Score: 0.6796668251040214


#### In this case we randomly select alpha value as 1.0 and got the accuracy score on test dataset as 67% as compared to just 13% as was in Simple Linear Regression case.
However, we can use GridSearchCV to select the best value of alpha, which is found below:

In [82]:
from sklearn.linear_model import Ridge
#Grid search is an approach to parameter tuning that will methodically build and evaluate a model for each combination of algorithm parameters specified in a grid.
from sklearn.model_selection import GridSearchCV
ridge=Ridge()
#Here alpha is lambda: is the parameter which balances the amount of emphasis given to minimizing RSS vs minimizing sum of square of coefficients
#These values of alpha have been chosen so that we can easily analyze the trend with change. These would however differ from case to case.
parameters={'alpha':[1e-15, 1e-10, 1e-8, 1e-3, 1e-2, 1, 5, 10, 20, 30, 35, 40, 45, 50, 55, 100]}
ridge_regressor=GridSearchCV(ridge,parameters,scoring='neg_mean_squared_error',cv=5) 
ridge_regressor.fit(X_train,y_train)
#Shows the best value of alpha that fits model
print(ridge_regressor.best_params_)
#Greater the value better the results
print(ridge_regressor.best_score_)

c:\Users\Aims tech\anaconda3\lib\site-packages\sklearn\linear_model\_ridge.py:157: LinAlgWarning: Ill-conditioned matrix (rcond=1.14828e-22): result may not be accurate.
  return linalg.solve(A, Xy, sym_pos=True, overwrite_a=True).T
c:\Users\Aims tech\anaconda3\lib\site-packages\sklearn\linear_model\_ridge.py:157: LinAlgWarning: Ill-conditioned matrix (rcond=1.14516e-22): result may not be accurate.
  return linalg.solve(A, Xy, sym_pos=True, overwrite_a=True).T
c:\Users\Aims tech\anaconda3\lib\site-packages\sklearn\linear_model\_ridge.py:157: LinAlgWarning: Ill-conditioned matrix (rcond=1.14151e-22): result may not be accurate.
  return linalg.solve(A, Xy, sym_pos=True, overwrite_a=True).T
c:\Users\Aims tech\anaconda3\lib\site-packages\sklearn\linear_model\_ridge.py:157: LinAlgWarning: Ill-conditioned matrix (rcond=1.14644e-22): result may not be accurate.
  return linalg.solve(A, Xy, sym_pos=True, overwrite_a=True).T
c:\Users\Aims tech\anaconda3\lib\site-packages\sklearn\linear_model\

{'alpha': 5}
-160505890782.9135


We have gotton the alpha value as 5.
So, trying with alpha = 5

In [83]:
from sklearn.linear_model import Ridge
l2 = Ridge(alpha=5)
l2.fit(X_train, y_train)
print('Ridge Test Score:',l2.score(X_test, y_test))
print('Ridge Train Score:',l2.score(X_train, y_train))

Ridge Test Score: 0.6743855819360092
Ridge Train Score: 0.6768851181317722


Here we got a very little increase in accuracy score on dataset.

### Lasso Regression
Lasso
Cost function = Loss + λ x ∑‖w‖ 

where,
Loss = Sum of the squared residuals (SSE), λ = Penalty for the errors, w = slope of the curve/line


LASSO stands for Least Absolute Shrinkage and Selection Operator. There are 2 key words here – ‘absolute‘ and ‘selection‘.
With Lasso regression, it’s possible that some of the coefficients could go completely to zero when λ i.e. α gets sufficiently large.
LASSO regression goes a bit extreme. It sets some feature coefficients to zero through L1 Regularization. This process essentially eliminates those features from the model instead of minimizing their impacts.
- Performs L1 regularization, i.e. adds penalty equivalent to absolute value of the magnitude of coefficients
- Minimization objective = LS Obj + α * (sum of absolute value of coefficients)

Lasso regression performs L1 regularization, i.e. it adds a factor of sum of absolute value of coefficients in the optimization objective. Thus, lasso regression optimizes the following:

Objective = RSS + α * (sum of absolute value of coefficients)
Here, α (alpha) provides a trade-off between balancing RSS and magnitude of coefficients i.e. α can take various values. 

Lets iterate it here briefly:
- α = 0: Same coefficients as simple linear regression
- α = ∞: All coefficients zero (same logic as before)
- 0 < α < ∞: coefficients between 0 and that of simple linear regression

In [84]:
from sklearn.linear_model import Lasso
l1 = Lasso(alpha=1)
l1.fit(X_train, y_train)
print('Lasso Test Score:', l1.score(X_test, y_test))
print('Lasso Train Score:', l1.score(X_train, y_train))

Lasso Test Score: 0.15711785249109578
Lasso Train Score: 0.6827631970269978


c:\Users\Aims tech\anaconda3\lib\site-packages\sklearn\linear_model\_coordinate_descent.py:647: ConvergenceWarning: Objective did not converge. You might want to increase the number of iterations, check the scale of the features or consider increasing regularisation. Duality gap: 1.265e+15, tolerance: 7.963e+11
  model = cd_fast.enet_coordinate_descent(


Here, we got just 15.71% accuracy score on test dataset.

Therefore, trying bigger value, we get

In [85]:
from sklearn.linear_model import Lasso
l1 = Lasso(alpha=5)
l1.fit(X_train, y_train)
print('Lasso Test Score:', l1.score(X_test, y_test))
print('Lasso Train Score:', l1.score(X_train, y_train))

Lasso Test Score: 0.22783016674029055
Lasso Train Score: 0.682572047831017


c:\Users\Aims tech\anaconda3\lib\site-packages\sklearn\linear_model\_coordinate_descent.py:647: ConvergenceWarning: Objective did not converge. You might want to increase the number of iterations, check the scale of the features or consider increasing regularisation. Duality gap: 1.273e+15, tolerance: 7.963e+11
  model = cd_fast.enet_coordinate_descent(


Again, we got just 22.78% accuracy score on test dataset.

Therefore, trying bigger value, we get

In [86]:
from sklearn.linear_model import Lasso
l1 = Lasso(alpha=50)
l1.fit(X_train, y_train)
print('Lasso Test Score:', l1.score(X_test, y_test))
print('Lasso Train Score:', l1.score(X_train, y_train))

Lasso Test Score: 0.6636263035287673
Lasso Train Score: 0.6767138609484034


c:\Users\Aims tech\anaconda3\lib\site-packages\sklearn\linear_model\_coordinate_descent.py:647: ConvergenceWarning: Objective did not converge. You might want to increase the number of iterations, check the scale of the features or consider increasing regularisation. Duality gap: 7.086e+14, tolerance: 7.963e+11
  model = cd_fast.enet_coordinate_descent(


Now, we have gotton some compareable results as in case of Ridge Regression were received. 

### L2 vs L1 Regularization

Where L1 regularization attempts to estimate the median of data, L2 regularization makes estimation for the mean of the data in order to evade overfitting.

Through including the absolute value of weight parameters, L1 regularization can add the penalty term in cost function. On the other hand, L2 regularization appends the squared value of weights in the cost function.

As defined, sparsity is the characteristic of holding highly significant coefficients, either very close to zero or not very close to zero, where in general coefficients approaching zero would be eliminated later. 

And the feature selection is the in-depth of sparsity, i.e. in place of confining coefficients nearby to zero, feature selection is brought them exactly to zero, and hence expel certain features from the data model. 

In this context, L1 regularization can be helpful in features selection by eradicating the unimportant features, whereas, L2 regularization is not recommended for feature selection.

L2 has a solution in closed form as it’s a square of a weight, on the other side, L1 doesn’t have a closed form solution since it includes an absolute value and it is a non-differentiable function. 

Due to this reason, L1 regularization is relatively more expensive in computation, it can’t be solved in the context of matrix measurement and heavily relies on approximations. 
L2 regularization is likely to be more accurate in all the circumstances, however, at a much higher level of computational costs.
### Drawback
The drawback of ridge and lasso regression is that it becomes difficult to interpret the coefficients in the final model since they get shrunk towards zero. Thus, ridge and lasso regression should be used when you’re interested in optimizing for predictive ability rather than inference.

### ElasticNet
Using both L1 and L2 regularization in the same regression. When we use both regularization terms in the model, it is called ElasticNet Regression.

### Conclusion
 

In order to prevent overfitting, regularization is most-approaches mathematical technique, it achieves this by panelizing the complex ML models via adding regularization terms to the loss function/cost function of the model. 

 

L1 regularization gives output in binary weights from 0 to 1 for the model’s features and is adopted for decreasing the number of features in a huge dimensional dataset. 

L2 regularization disperse the error terms in all the weights that leads to more accurate customized final models.